srml/contract/src/wasm/code.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate. If not, see <http://www.gnu.org/licenses/>.
+
+use gas::{GasMeter, Token};
+use rstd::prelude::*;
+use runtime_primitives::traits::{As, CheckedMul, Hash};
+use runtime_support::StorageMap;
+use wasm::{prepare, runtime::Env, PrefabWasmModule};
+use {CodeHash, CodeStorage, PristineCode, Schedule, Trait};
+
+/// Gas metering token that used for charging storing code into the code storage.
+///
+/// Specifies the code length in bytes.
+#[cfg_attr(test, derive(Debug, PartialEq, Eq))]
+#[derive(Copy, Clone)]
+pub struct PutCodeToken(u64);
+
+impl<T: Trait> Token<T> for PutCodeToken {
+	type Metadata = Schedule<T::Gas>;
+
+	fn calculate_amount(&self, metadata: &Schedule<T::Gas>) -> Option<T::Gas> {
+		let code_len_in_gas = <T::Gas as As<u64>>::sa(self.0);
+		let cost = metadata
+			.put_code_per_byte_cost
+			.checked_mul(&code_len_in_gas)?;
+
+		Some(cost)
+	}
+}
+
+/// Put code in the storage. The hash of code is used as a key and is returned
+/// as a result of this function.
+///
+/// This function instruments the given code and caches it in the storage.
+pub fn save<T: Trait>(
+	original_code: Vec<u8>,
+	gas_meter: &mut GasMeter<T>,
+	schedule: &Schedule<T::Gas>,
+) -> Result<CodeHash<T>, &'static str> {
+	// The first time instrumentation is on the user. However, consequent reinstrumentation
+	// due to the schedule changes is on governance system.
+	if gas_meter
+		.charge(schedule, PutCodeToken(original_code.len() as u64))
+		.is_out_of_gas()
+	{
+		return Err("there is not enough gas for storing the code");
+	}
+
+	let prefab_module = prepare::prepare_contract::<T, Env>(&original_code, schedule)?;
+	let code_hash = T::Hashing::hash(&original_code);
+
+	// TODO: validate the code. If the code is not valid, then don't store it.
```

Comment:
Another great starter issue I see right here!

Reply:
Totally agree! It is blocked on changes to wasmi and/or parity-wasm and they are already taken by Fedor!

Reply:
Though you're right, will create one